cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh

```diff
@@ -0,0 +1,1043 @@
+  }
+
+  //
+  // Make ext_indices
+  //
+  for (uint32_t l = 0; l < ext_index.n_lists(); l++) {
+    auto old_cluster_size =
+      static_cast<uint32_t>(old_cluster_offsets[l + 1] - old_cluster_offsets[l]);
+    auto new_cluster_size = cluster_sizes.data()[l];
+    if (old_cluster_size > 0) {
+      copy(ext_indices + ext_cluster_offsets[l],
+           orig_index.indices().data_handle() + old_cluster_offsets[l],
+           old_cluster_size,
+           stream);
+    }
+    if (new_cluster_size > 0) {
+      if (new_indices == nullptr) {
+        // implies the orig index is empty
+        copy(ext_indices + ext_cluster_offsets[l] + old_cluster_size,
+             data_indices.data() + cluster_offsets[l],
+             new_cluster_size,
+             stream);
+      } else {
+        utils::copy_selected(new_cluster_size,
+                             1,
+                             new_indices,
+                             data_indices.data() + cluster_offsets[l],
+                             1,
+                             ext_indices + ext_cluster_offsets[l] + old_cluster_size,
+                             1,
+                             stream);
+      }
+    }
+  }
+
+  /* Extend the pq_dataset */
+  auto ext_pq_dataset    = ext_index.pq_dataset().data_handle();
+  size_t pq_dataset_unit = ext_index.pq_dim() * ext_index.pq_bits() / 8;
+  for (uint32_t l = 0; l < ext_index.n_lists(); l++) {
+    auto old_cluster_size =
+      static_cast<uint32_t>(old_cluster_offsets[l + 1] - old_cluster_offsets[l]);
+    copy(ext_pq_dataset + pq_dataset_unit * ext_cluster_offsets[l],
+         orig_index.pq_dataset().data_handle() + pq_dataset_unit * old_cluster_offsets[l],
+         pq_dataset_unit * old_cluster_size,
+         stream);
+    copy(ext_pq_dataset + pq_dataset_unit * (ext_cluster_offsets[l] + old_cluster_size),
+         new_pq_codes.data() + pq_dataset_unit * cluster_offsets[l],
+         pq_dataset_unit * cluster_sizes.data()[l],
+         stream);
+  }
+  handle.sync_stream();
+
+  _cuann_get_inclusiveSumSortedClusterSize(ext_index);
+
+  return ext_index;
+}
+
+/** See raft::spatial::knn::ivf_pq::build docs */
+template <typename T, typename IdxT>
+inline auto build(
+  const handle_t& handle, const index_params& params, const T* dataset, IdxT n_rows, uint32_t dim)
+  -> index<IdxT>
+{
+  common::nvtx::range<common::nvtx::domain::raft> fun_scope(
+    "ivf_pq::build(%zu, %u)", size_t(n_rows), dim);
+  static_assert(std::is_same_v<T, float> || std::is_same_v<T, uint8_t> || std::is_same_v<T, int8_t>,
+                "Unsupported data type");
+
+  RAFT_EXPECTS(n_rows > 0 && dim > 0, "empty dataset");
+
+  auto stream = handle.get_stream();
+
+  ivf_pq::index<IdxT> index(handle, params, dim);
+  utils::memzero(index.list_offsets().data_handle(), index.list_offsets().size(), stream);
+  utils::memzero(index.pq_centers().data_handle(), index.pq_centers().size(), stream);
+
+  auto trainset_ratio = std::max<IdxT>(
```

Comment:
There's slight difference in that ivf-pq also translates the input type into float while sampling. I'm thinking to put this into a follow-up pr later, because the sampling code needs the change anyway (https://github.com/rapidsai/raft/issues/711 (5)).

Reply:
Ok, that make sense.